lib/rspec/matchers/built_in/yield.rb

```diff
@@ -369,16 +363,18 @@ def initialize(*args)
         # @private
         def matches?(block)
           @actual_formatted = []
+          @actual = []
           args_matched_when_yielded = true
           yield_count = 0
 
-          @probe = YieldProbe.probe(block) do |args|
-            @actual_formatted << RSpec::Support::ObjectFormatter.format(args)
-            args_matched_when_yielded &&= values_match?(@expected[yield_count], args)
+          @probe = YieldProbe.probe(block) do |*arg_array|
```

Comment:
The fix has two parts to it.  After the changes in #965, this block only accepted a single, non-splatted arg.  Ruby's semantics dictate that that makes the extra arguments get dropped on the floor.  As a result, it only compared the expected values against the _first_ yielded argument of each yield.

The splat is necessary so that we receive _all_ the arguments on each yield, so we can compare them to what is expected.  However, doing that is not enough.  It fixed the new specs I added but broke most (or all) of the other specs for this matcher.

That's because the splat causes the `arg_array` variable to _always_ be an array, even if only one value was yielded (in which case, it's an array of one value).  The `successive_yield_args` thing that I inlined was what was handling that before, but w/o inlining the logic, it was only involved in determining `@actual` (since that's where it was called from) and not actually involved in the match logic.  By in-lining that method's logic, it ensures that if only one value is yielded, we deal with individual values, but if multiple values are yielded, we deal with an array for each yield.

Does that make sense?